### PR TITLE
[Unified checkpoint] Add unified optimizer

### PR DESCRIPTION
**Pull request by DesmonDay**
### PR types
New features

### PR changes
Others

### Description

Add unified optimizer which merge tp tensor, and save all optimizers into `optimizer.safetensors.index.json` and `master_weights.safetensors.index.json` if exists.

- eg. for tp4pp2, 
- it will save `optimizer-00001-of-00008.safetensors ~ optimizer-00008-of-00008.safetensors`
- it will save `master_weights-00001-of-00008.safetensors ~ master_weights-00008-of-00008.safetensors` if exists
- `00001-00004` store all tensor for pp00, `00005-00008` store all tensor for pp01
-  each of `00001-00004`  are tensor merge on tensor parallel, and equally split into 4 parts.

#### Current save state
``` python
-rw-r--r-- 1 root root 1015 Nov 21 16:40 config.json
-rw-r--r-- 1 root root 3.1G Nov 21 16:40 master_weights-00001-of-00008.safetensors
-rw-r--r-- 1 root root 3.2G Nov 21 16:40 master_weights-00002-of-00008.safetensors
-rw-r--r-- 1 root root 3.2G Nov 21 16:40 master_weights-00003-of-00008.safetensors
-rw-r--r-- 1 root root 3.2G Nov 21 16:40 master_weights-00004-of-00008.safetensors
-rw-r--r-- 1 root root 3.1G Nov 21 16:40 master_weights-00005-of-00008.safetensors
-rw-r--r-- 1 root root 3.2G Nov 21 16:40 master_weights-00006-of-00008.safetensors
-rw-r--r-- 1 root root 3.1G Nov 21 16:40 master_weights-00007-of-00008.safetensors
-rw-r--r-- 1 root root 3.3G Nov 21 16:40 master_weights-00008-of-00008.safetensors
-rw-r--r-- 1 root root  28K Nov 21 16:40 master_weights.safetensors.index.json
-rw-r--r-- 1 root root 1.6G Nov 21 16:40 model-00001-of-00008.safetensors
-rw-r--r-- 1 root root 1.6G Nov 21 16:40 model-00002-of-00008.safetensors
-rw-r--r-- 1 root root 1.6G Nov 21 16:40 model-00003-of-00008.safetensors
-rw-r--r-- 1 root root 1.6G Nov 21 16:40 model-00004-of-00008.safetensors
-rw-r--r-- 1 root root 1.6G Nov 21 16:40 model-00005-of-00008.safetensors
-rw-r--r-- 1 root root 1.6G Nov 21 16:40 model-00006-of-00008.safetensors
-rw-r--r-- 1 root root 1.6G Nov 21 16:40 model-00007-of-00008.safetensors
-rw-r--r-- 1 root root 1.7G Nov 21 16:40 model-00008-of-00008.safetensors
-rw-r--r-- 1 root root  25K Nov 21 16:40 model.safetensors.index.json
-rw-r--r-- 1 root root 6.2G Nov 21 16:40 optimizer-00001-of-00008.safetensors
-rw-r--r-- 1 root root 6.4G Nov 21 16:40 optimizer-00002-of-00008.safetensors
-rw-r--r-- 1 root root 6.2G Nov 21 16:40 optimizer-00003-of-00008.safetensors
-rw-r--r-- 1 root root 6.4G Nov 21 16:40 optimizer-00004-of-00008.safetensors
-rw-r--r-- 1 root root 6.3G Nov 21 16:40 optimizer-00005-of-00008.safetensors
-rw-r--r-- 1 root root 6.4G Nov 21 16:40 optimizer-00006-of-00008.safetensors
-rw-r--r-- 1 root root 6.3G Nov 21 16:40 optimizer-00007-of-00008.safetensors
-rw-r--r-- 1 root root 6.4G Nov 21 16:40 optimizer-00008-of-00008.safetensors
-rw-r--r-- 1 root root 118K Nov 21 16:40 optimizer.safetensors.index.json
-rw-r--r-- 1 root root  55K Nov 21 16:40 rng_state_8.pth
-rw-r--r-- 1 root root  317 Nov 21 16:40 scaler.pdparams
-rw-r--r-- 1 root root   50 Nov 21 16:40 scheduler.pdparams
-rw-r--r-- 1 root root 489K Nov 21 16:40 sentencepiece.bpe.model
-rw-r--r-- 1 root root   63 Nov 21 16:40 special_tokens_map.json
-rw-r--r-- 1 root root  207 Nov 21 16:40 tokenizer_config.json
-rw-r--r-- 1 root root  596 Nov 21 16:40 trainer_state.json
-rw-r--r-- 1 root root 2.2K Nov 21 16:40 training_args.bin
```

#### Optimizer name design

To improve the readability of optimizer parameters, we renamed the keys of the optimizer parameters. The original optimizer parameters mainly contains static name, for example:
- `embedding_0.w_0_fp32_master_0_moment1_0`
- `embedding_0.w_0_fp32_master_0_beta1_pow_acc_0`
- `linear_0.w_0_fp32_master0_moment1_0`

Instead, we get the structure name and specific tensor name, and build the new optimizer key names, for example:
- `embedding.tokens.weight/moment1_0`
- `embedding.tokens.weight/beta1_pow_acc_0`
- `llama.layers.0.self_attn.q_proj.weight/moment1_0`

Here we paste part of `optimizer.safetensors.index.json`:
``` 
{
    "metadata": {
        "total_size": 53907327256
    },   
    "weight_map": {
        "llama.embed_tokens.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.0.self_attn.q_proj.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.0.self_attn.k_proj.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.0.self_attn.v_proj.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.0.self_attn.o_proj.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.0.mlp.gate_proj.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.0.mlp.up_proj.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.0.mlp.down_proj.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.0.input_layernorm.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.0.post_attention_layernorm.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.1.self_attn.q_proj.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.1.self_attn.k_proj.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.1.self_attn.v_proj.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.1.self_attn.o_proj.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.1.mlp.gate_proj.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.1.mlp.up_proj.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.1.mlp.down_proj.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.1.input_layernorm.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        "llama.layers.1.post_attention_layernorm.weight/moment1_0": "optimizer-00001-of-00008.safetensors",
        ...
    },
   "master_weights": true
}
```

### CODE REVIEW

**Comment by DesmonDay** [comment]
由于将 concat操作放在GPU上，会导致显存增加，一般为最大完整Tensor的显存。在for循环中可增加判断，计算某个tensor大小，如超过某个阈值，则采用offload方式，否则仍按照GPU方式进行concat。

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 1]
```diff
@@ -399,3 +665,55 @@ def merge_tensor_parallel_with_shard(model_to_save, state_dict, config, all_filt
...
+            model_key = static2struct_name_mappings[
+                filter_keys[i] if is_master_weights else generate_base_static_name(filter_keys[i])[0]
+            ]
+            tensor = state_dict[filter_keys[i]]
+            if model_key in tp_actions:
+                # for example: beta1, beta2
+                if tensor.numel().item() == 1:
+                    tensor = (
+                        tensor._copy_to(paddle.CPUPlace(), False) if is_dst else None
+                    )  # Need broadcast when loaded
```
Comment (DesmonDay) [inline paddlenlp/trainer/plugins/unified_checkpoint.py:708]: 是broadcast还是单纯send需要具体确定下

File: paddlenlp/transformers/conversion_utils.py [hunk 2]
```diff
@@ -336,7 +336,7 @@ def naive_fuse_split_tp(
     return np.concatenate(splited[tensor_parallel_rank::tensor_parallel_degree], axis=axis)
 
 
-def normal_fuse_merge_tp(weight_list, is_column=True):
+def normal_fuse_merge_tp(weight_list, is_column=True, return_numpy=True):
```
Comment (ZHUI) [inline paddlenlp/transformers/conversion_utils.py]: 不要加参数，直接判断 `weight_list` 变量类型。

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 3]
```diff
@@ -399,3 +665,55 @@ def merge_tensor_parallel_with_shard(model_to_save, state_dict, config, all_filt
...
+                else:
+                    ret = distributed_gather(tensor, dst=j, group=tp_group, offload=False)
+                    action = partial(tp_actions[model_key], return_numpy=False)
```
Comment (ZHUI) [inline paddlenlp/trainer/plugins/unified_checkpoint.py]: 删除 `return_numpy`

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 4]
```diff
@@ -266,36 +486,83 @@ def unified_checkpoint_into_shards(
...
+
+    if tp_rank == 0:
+        tensor_bytes_dict = {}
+        model_state_dict = model_to_save.state_dict()
+        for (k, v) in state_dict.items():
+            if is_master_weights:
+                model_v = model_state_dict[static2struct_name_mappings[k]]
+            else:
+                model_v = model_state_dict[static2struct_name_mappings[generate_base_static_name(k)[0]]]
+            if hasattr(model_v, "is_distributed") and model_v.is_distributed:
```
Comment (DesmonDay) [inline paddlenlp/trainer/plugins/unified_checkpoint.py:533]: 这块对于标量值，比如beta，也会乘以 tp_size。看是否要修改。

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 5]
```diff
@@ -50,9 +48,21 @@
     from safetensors.numpy import save_file as safe_save_file
 
 
+PADDLE_OPTIMIZER_NAME = "optimizer.pdopt"
+PADDLE_OPTIMIZER_INDEX_NAME = "optimizer.pdopt.index.json"
+SAFE_OPTIMIZER_NAME = "optimizer.safetensors"
+SAFE_OPTIMIZER_INDEX_NAME = "optimizer.safetensors.index.json"
+PADDLE_MASTER_WEIGHTS_NAME = "master_weights.pdopt"
+PADDLE_MASTER_WEIGHTS_INDEX_NAME = "master_weights.pdopt.index.json"
```
Comment (ZHUI) [inline paddlenlp/trainer/plugins/unified_checkpoint.py]:
```suggestion
PADDLE_MASTER_WEIGHTS_INDEX_NAME = "master_weights.pdparams.index.json"
```

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 6]
```diff
@@ -50,9 +48,21 @@
     from safetensors.numpy import save_file as safe_save_file
 
 
+PADDLE_OPTIMIZER_NAME = "optimizer.pdopt"
+PADDLE_OPTIMIZER_INDEX_NAME = "optimizer.pdopt.index.json"
+SAFE_OPTIMIZER_NAME = "optimizer.safetensors"
+SAFE_OPTIMIZER_INDEX_NAME = "optimizer.safetensors.index.json"
+PADDLE_MASTER_WEIGHTS_NAME = "master_weights.pdopt"
```
Comment (ZHUI) [inline paddlenlp/trainer/plugins/unified_checkpoint.py]:
```suggestion
PADDLE_MASTER_WEIGHTS_NAME = "master_weights.pdparms"
```

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 7]
```diff
@@ -221,38 +231,232 @@ def unified_checkpoint_into_shards(
...
+
+    Args:
+        optimizer (Optimizer): optimizer to save.
+        safe_serialization (bool, optional): safe serialization using safetensors. Defaults to False.
+    """
+    optim_state_dict = nested_copy(optimizer.state_dict())
+    master_weights = None
+    if "master_weights" in optim_state_dict.keys():
+        master_weights = optim_state_dict["master_weights"]
+    optim_state_dict.pop("master_weights")
```
Comment (ZHUI) [inline paddlenlp/trainer/plugins/unified_checkpoint.py]: 没有`master_weights`的话，会不会报错？

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 8]
```diff
@@ -221,38 +231,232 @@ def unified_checkpoint_into_shards(
...
+    optim_state_dict.pop("LR_Scheduler")
```
Comment (ZHUI) [inline paddlenlp/trainer/plugins/unified_checkpoint.py]: 注意报错

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 9]
```diff
@@ -399,3 +589,58 @@ def merge_tensor_parallel_with_shard(model_to_save, state_dict, config, all_filt
             logger.warning(f"key <{x}> need to merge tensor parallel but we can't find in model state.")
 
     return state_dict_to_save
+
+
+def merge_tensor_parallel_for_optimizer(model_to_save, state_dict, config, all_filter_keys):
```
Comment (ZHUI) [inline paddlenlp/trainer/plugins/unified_checkpoint.py]: 这个函数，直接传 `tp_actions` 到这个函数就好了，不需要 `model_to_save`。简单一些。

Comment (ZHUI) [reply paddlenlp/trainer/plugins/unified_checkpoint.py]: config好像也不要

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 10]
```diff
@@ -399,3 +589,58 @@ def merge_tensor_parallel_with_shard(model_to_save, state_dict, config, all_filt
...
+    for i in range(max_key_len):
+        for j, filter_keys in enumerate(all_filter_keys):
+            is_dst = tp_rank == j
+            if i > len(filter_keys) - 1:
+                continue
+            # get base model key
+            model_key = filter_keys[i].split("/")[0]
+            tensor = state_dict[filter_keys[i]]
+            if model_key in tp_actions:
+                # for example: beta1, beta2
```
Comment (ZHUI) [inline paddlenlp/trainer/plugins/unified_checkpoint.py:631]: too hack for readers

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 11]
```diff
@@ -399,3 +589,58 @@ def merge_tensor_parallel_with_shard(model_to_save, state_dict, config, all_filt
...
+    tp_actions = model_to_save.get_tensor_parallel_convert_actions(
+        model_to_save.config, model_keys, is_split=False, ignore_error=True
+    )
+
+    hcg = fleet.get_hybrid_communicate_group()
+    tp_group = hcg.get_model_parallel_group()
+    tp_rank = tp_group.rank
+
+    state_dict_to_save = {}
+    max_key_len = max([len(_) for _ in all_filter_keys])
```
Comment (ZHUI) [inline paddlenlp/trainer/plugins/unified_checkpoint.py:621]: shuffle一下，应该比你这样按照key的名字排序也要好一点吧。

File: paddlenlp/trainer/plugins/unified_checkpoint.py [hunk 12]
```diff
@@ -221,38 +231,232 @@ def unified_checkpoint_into_shards(
...
+
+    # filter optimizer param
+    if master_weights is not None:
+        filter_master_keys = filter_params(model, master_weights, is_optimizer=True)
+    filter_optim_keys = filter_params(model, optim_state_dict, is_optimizer=True)
+
+    tp_group = fleet.get_hybrid_communicate_group().get_model_parallel_group()
+    tp_size = tp_group.nranks
+
+    if tp_size > 1:
```
Comment (ZHUI) [inline paddlenlp/trainer/plugins/unified_checkpoint.py:359]: 注意验证非tp的情况

File: paddlenlp/trainer/plugins/unified_checkpoint.py (hunk 11 above)
Comment (DesmonDay) [reply paddlenlp/trainer/plugins/unified_checkpoint.py:621]: 如果使用shuffle，还需另外维护一份keys对应需要的rank，个人感觉没有太大必要。现在的写法也可以保证每个rank并发接收需要的tensor，满足需求。

File: paddlenlp/trainer/plugins/unified_checkpoint.py (hunk 7 above)
Comment (DesmonDay) [reply paddlenlp/trainer/plugins/unified_checkpoint.py]: 已修改

File: paddlenlp/trainer/plugins/unified_checkpoint.py (hunk 8 above)
Comment (DesmonDay) [reply paddlenlp/trainer/plugins/unified_checkpoint.py]: done

File: paddlenlp/trainer/plugins/unified_checkpoint.py (hunk 4 above)
Comment (ZHUI) [reply paddlenlp/trainer/plugins/unified_checkpoint.py:533]: 后面优化

**Review by ZHUI — APPROVED** [review]
LGTM

**Review by sijunhe — APPROVED** [review]

**Comment by ZHUI** [comment]
### model不保存
1. fp32 的参数
2. stop_gradient

### 3种情况
1. 存model
2. 不存model
4. 只存fp32 model
